### ISSUE TITLE
[BUG] If we overlap line masking, Notepad++ freezes/crashes.

### ISSUE THREAD

**Issue by Snowball-V** [notepad-plus-plus/notepad-plus-plus#15630]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description of the Issue

When we hide already hidden lines, Notepad++ freezes/crash. It's not systematic, and I mostly feel that the issue occurs when the **End of File (EoF) is included** in the hidden section.

### Steps To Reproduce

---

1/ Download a Notepad++ Portable Minimalist version to ensure there are no unwanted plugins.

---

2/ Extract it.

---

3/ Launch Notepad++

---

4/ Write the following lines:
```
{
	// sample
	// sample
}
{
	// sample
	// sample
}
```
<details>
<summary>Image</summary>

![Description de l'image](https://i.ibb.co/db9p82z/15-09-2024-12-52-12.png)

</details>

---

5/ Highlight the last 4 lines (lines 8 to 5) and press `ALT + H` to hide them.
<details>
<summary>Image</summary>

![Description de l'image](https://i.ibb.co/rp6RYb8/15-09-2024-12-52-17.png)

</details>

---

6/ Now, you should only see this:
```
{
	// sample
	// sample
}
}
```
With the fold markers on line 4 (>) and line 8 (<).

<details>
<summary>Image</summary>

![Description de l'image](https://i.ibb.co/ft1P5hN/15-09-2024-12-52-22.png)

</details>

---

7/ Highlight all 5 lines (lines 8 to 1, including lines 5, 6, and 7, which are hidden) and press `ALT + H` to hide them.

<details>
<summary>Image</summary>

![Description de l'image](https://i.ibb.co/mF4YyPc/15-09-2024-13-18-43.png)

</details>

---

8/ Notepad++ freezes.

<details>
<summary>Image</summary>

![Description de l'image](https://i.ibb.co/BqZ1tCn/15-09-2024-12-52-44.png)

</details>

---

### Current Behavior

Notepad++ Freeze.

### Expected Behavior

I expect the lines to hide and unhide normally while respecting previous fixes that ensure hidden lines remain hidden when folding or unfolding.

I'm a beginner programmer, and I acknowledge that understanding the logic behind what should be hidden or not, keeping track of what needs to remain hidden, and managing overlapping 'un/hide line' and 'un/fold line' operations can be quite a headache.

I would fully understand and accept that this bug is not fixable in order to avoid breaking the rest of Notepad++'s code.

### Debug Information

```shell
Notepad++ v8.6.9   (64-bit)
Build time : Jul 12 2024 - 05:09:25
Path : C:\Downloads\npp.8.6.9.portable.minimalist.x64\notepad++.exe
Command Line : C:\Downloads\npp.8.6.9.portable.minimalist.x64\bug.txt
Admin mode : OFF
Local Conf mode : ON
Cloud Config : OFF
Periodic Backup : ON
OS Name : Windows 8.1 (64-bit)
OS Build : 9600.20045
Current ANSI codepage : 1252
Plugins : none
```

### Anything else?

Related issues ?

https://github.com/notepad-plus-plus/notepad-plus-plus/issues/13682
https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12184

**Comment by alankilborn** [comment]
Unfortunately, hiding of lines is a very immature feature in Notepad++.  I'd like to use it, but I've found that it is so immature that any non-trivial use is pointless.

**Comment by xomx** [comment]
Infinite loop here (`scope` will never be 0):
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/45cd29f8ea700b602550e5329bac8ab90b341313/PowerEditor/src/ScintillaComponent/ScintillaEditView.cpp#L3997-L3998

@Adrian104 
Can you adapt your code to fix it?

**Comment by Adrian104** [comment]
Sure, I'll try to fix that

**Comment by Snowball-V** [comment]
This message won't help you, but I admit that I have my share of responsibility because as I progress in my code, I tend to fold, refold, and re-re-re-refold the lines above somewhat blindly without checking if I’m overlapping existing folds…

Naturally, Notepad++ gets a bit overwhelmed with this usage.

My issue is that since these freezes or crashes, I lose all unsaved data.

And since I understand that it's an issue with an infinite loop, maybe we could add a fallback or counter that, if it detects more than few million lines being folded, it recognizes there's probably an issue and exits the loop early and cleanly to give control back to the user ? At the very least, this would give time to save before freezing...

Note : I've noticed that it also happens in the middle of the page, so **it's not just an issue with folding at the end of the file** but rather with fold upon fold upon fold…

I'll try to find a way to reliably reproduce the bug to help with the investigation.

_Thank you all so much for making Notepad++, and thus the world, a better place with your skilled PR and contributions !_

**Comment by Adrian104** [comment]
It should be fixed now. In a nutshell, there was a problem when removing old markers before creating new ones: program removed marker on line 4, skipped 8, and tried to delete markers past 8th line. This caused an infinity loop.

@Snowball-V @xomx 
Could you please verify that after PR #15632 hiding lines works properly? Of course, I have tested it myself, but I don't want to miss some other edge cases like that again. Thank you in advance!

**Comment by Snowball-V** [comment]
I'm not quite sure how to test it, so I simply copied your PR / modification into the Master source code and compiled Notepad++ from this edited source with your change, and… I can't reproduce the bug described in this message. 👏

So, I'll go ahead and use this edited/compiled version in daily use to see if the issue pops up unexpectedly. As long as I don't give any updates here, it means this fix is holding up!

Thank you very much @Adrian104 for your quick response and for the fix! 👌

**Comment by Adrian104** [comment]
> Edit : I managed to make it freeze again despite the fix, but it's much rarer now, so I need to figure out how to reproduce the issue consistently to have a detailed description for explaining.

I have found a way. First, copy following lines and paste them into Notepad++:
```

1. Hide
3. Delete
2. Hide
   Also delete
4. Hide

```

Then follow the instructions from 1 to 4, in order. Demonstration:
![freeze](https://github.com/user-attachments/assets/c8832d26-2280-43cb-b832-6ac62fae59e2)

After debugging for a while, it seems like this problem is caused by deleting hidden lines (or adjacent lines to them). After deletion, markers may enter an abnormal state, where the number of "begin" markers does not equal to the number of "end" markers. This invalid state is then read by `hideLines()` function, which crashes the program.

However, this bug seems to be unrelated to the `hideLines()` function, because invalid state begins somewhere else. In other words, freezing `hideLines()` is a side effect of some other problem(s).

**Comment by donho** [comment]
@Adrian104 
> After debugging for a while, it seems like this problem is caused by deleting hidden lines (or adjacent lines to them). After deletion, markers may enter an abnormal state, where the number of "begin" markers does not equal to the number of "end" markers. This invalid state is then read by hideLines() function, which crashes the program.

With your example I can reproduce the hanging, and found the following responsible lines which makes infinit loop:

```cpp
		while (scope > 0)
			removeMarker(++endMarker, 1 << MARK_HIDELINESEND);
```
Even the marker's abnormal state, it's not reasonable to keep these infint loop. 
Controling the `scope` variable is not enough to stop the loop. the `endMarker` should be checked as well.

**Comment by Snowball-V** [comment]
> @Snowball-V, Just following up. It’s been almost two weeks, and I was wondering if you had any updates. Did this freeze happen more than once after the fix, or it was one-off? What action caused this to happen (hiding, unhiding or something else)? Thank you in advance!

Sorry for the delay, in the meantime it seems that the problem has been solved by 5d2bea5 with version 8.7.1.

Thanks donho and Adrian104 for your help and correction ! 👍😘
